File: Tree/Leetcode-1038-BinarySearchTreeToGreaterSumTree/Leetcode-1038-BinarySearchTreeToGreaterSumTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	TreeNode* bstToGst(TreeNode* root) {
		vector<int> valList;
		int index = 0;
		recursiveList(valList, root);
		for (int i = 1; i < valList.size(); ++i) {
			valList[i] += valList[i - 1];
		}
		recursiveChange(valList, root, index);

		return root;
	}

private:
	void recursiveList(vector<int>& valList, TreeNode* node) {
		if( node->right != 0 )
			recursiveList(valList, node->right);
		valList.push_back(node->val);
		if( node->left != 0 )
			recursiveList(valList, node->left);
	}
    
    void recursiveChange(vector<int>& valList, TreeNode* node, int& index) {
		if( node->right != 0 )
			recursiveChange(valList, node->right, index);
		node->val = valList[index++];
		if( node->left != 0 )
			recursiveChange(valList, node->left, index);
	}
};

File: 2024상반기/3월1주차/leetcode-1038-Binary Search Tree to Greater Sum Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        dq = collections.deque([root]) # bfs방식으로 tree를 array로 변환
        array = []
        while dq:
            for i in range(len(dq)):
                if dq[0] == None:
                    array.append(dq.popleft())
                else:
                    array.append(dq[0].val)
                    dq.append(dq[0].left)
                    dq.append(dq[0].right)
                    dq.popleft()
        
        modified = collections.deque(array)
        for i in range(len(array)):
            if array[i] == None:
                continue
            for j in range(len(array)):
                if array[j] == None:
                    continue
                if array[i] > array[j]:
                    modified[j] = modified[j] + array[i]

        dq = collections.deque([root])
        while dq:
            for i in range(len(dq)):
                node = dq.popleft()
                if not node:
                    modified.popleft()
                else:
                    node.val = modified.popleft()
                    dq.append(node.left)
                    dq.append(node.right)
        return root